synapse/lib/storm.py

```diff
@@ -1946,6 +1947,138 @@ async def execStormCmd(self, runt, genr):
+                    adds.clear()
+
+                if subs:
+                    subedits = [(node.buid, node.form.name, subs)]
+                    await runt.snap.view.layers[0].storNodeEdits(subedits, meta=meta)
+                    subs.clear()
+
+            # check all node perms first
+            form = sode.get('form')
+            valu = sode.get('valu')
+            if valu is not None:
+                runt.confirm(('node', 'add', form), gateiden=layr0)
+                adds.append((s_layer.EDIT_NODE_ADD, valu, ()))
+                subs.append((s_layer.EDIT_NODE_DEL, valu, ()))
+                if not self.opts.apply:
+                    valurepr = node.form.type.repr(valu)
+                    await runt.printf(f'{nodeiden} {form} = {valurepr}')
+
+            for name, (valu, stortype) in sode.get('props', {}).items():
+                full = node.form.prop(name).full
+                runt.confirm(('node', 'prop', 'del', full), gateiden=layr0)
+                runt.confirm(('node', 'prop', 'set', full), gateiden=layr1)
+                adds.append((s_layer.EDIT_PROP_SET, (name, valu, None, stortype), ()))
+                subs.append((s_layer.EDIT_PROP_DEL, (name, valu, stortype), ()))
+                if not self.opts.apply:
+                    valurepr = node.form.prop(name).type.repr(valu)
+                    await runt.printf(f'{nodeiden} {form}:{name} = {valurepr}')
+
+            for tag, valu in sode.get('tags', {}).items():
+                tagperm = tuple(tag.split('.'))
+                runt.confirm(('node', 'tag', 'del') + tagperm, gateiden=layr0)
+                runt.confirm(('node', 'tag', 'add') + tagperm, gateiden=layr1)
+                adds.append((s_layer.EDIT_TAG_SET, (tag, valu, None), ()))
+                subs.append((s_layer.EDIT_TAG_DEL, (tag, valu), ()))
+                if not self.opts.apply:
+                    valurepr = ''
+                    if valu != (None, None):
+                        tagrepr = runt.model.type('ival').repr(valu)
+                        valurepr = f' = {tagrepr}'
+                    await runt.printf(f'{nodeiden} {form}#{tag}{valurepr}')
+
+            for (tag, prop), (valu, stortype) in sode.get('tagprops', {}).items():
+                tagperm = tuple(tag.split('.'))
+                runt.confirm(('node', 'tag', 'del') + tagperm, gateiden=layr0)
+                runt.confirm(('node', 'tag', 'add') + tagperm, gateiden=layr1)
+                adds.append((s_layer.EDIT_TAGPROP_SET, (tag, prop, valu, None, stortype), ()))
+                subs.append((s_layer.EDIT_TAGPROP_DEL, (tag, prop, valu, stortype), ()))
+                if not self.opts.apply:
+                    valurepr = repr(valu)
+                    await runt.printf(f'{nodeiden} {form}#{tag}:{prop} = {valurepr}')
+
+            layr = runt.snap.view.layers[0]
+            async for name, valu in layr.iterNodeData(node.buid):
+                adds.append((s_layer.EDIT_NODEDATA_SET, (name, valu, None), ()))
```

Comment:
Should we check nodedata perms here or just assume that it's fine if they've had everything else up to here?